jsscripts/InputMethodHandler.js

```diff
@@ -0,0 +1,140 @@
+  // Weak-ref used to keep track of the currently focused element.
+  _currentFocusedElement: null,
+
+  _init: function() {
+    Logger.debug("JSScript: InputMethodHandler.js loaded");
+    addEventListener("focus", this, true);
+    addEventListener("blur", this, true);
+    addEventListener("input", this, false);
+  },
+
+  get focusedElement() {
+    return this._currentFocusedElement && this._currentFocusedElement.get();
+  },
+
+  notifySelectionChanged: function(doc, sel, reason) {
+    this._sendInputContext(this.focusedElement);
+  },
+
+  handleEvent: function(aEvent) {
+    switch (aEvent.type) {
+      case "focus": {
+        let currentElement = aEvent.target;
+        if (this._isAutoComplete(currentElement)) {
+          this._currentFocusedElement = Cu.getWeakReference(currentElement);
+          let selPriv = currentElement.editor.selectionController
+                        .getSelection(Ci.nsISelectionController.SELECTION_NORMAL)
+                        .QueryInterface(Ci.nsISelectionPrivate);
+          selPriv.addSelectionListener(this);
+        }
+        break;
+      }
+
+      case "blur": {
+        let focused = this.focusedElement;
+        if (focused) {
+          let selPriv = focused.editor.selectionController
+                        .getSelection(Ci.nsISelectionController.SELECTION_NORMAL)
+                        .QueryInterface(Ci.nsISelectionPrivate);
+          selPriv.removeSelectionListener(this);
+          this._resetInputContext(focused);
+        }
+        this._currentFocusedElement = null;
+        break;
+      }
+
+      case "input": {
+        let currentElement = aEvent.target;
+        if (currentElement !== this.focusedElement) {
+          break;
+        }
+        this._sendInputContext(currentElement);
+        break;
+      }
+    }
+  },
+
+  _sendInputContext: function(aElement) {
+    if (!this._isAutoComplete(aElement) || aElement !== this.focusedElement) {
+      return;
+    }
+
+    try {
+      let winId = Services.embedlite.getIDByWindow(aElement.ownerGlobal);
+      Services.embedlite.sendAsyncMessage(winId, "InputMethodHandler:SetInputContext",
+                                          JSON.stringify({surroundingText: aElement.value,
+                                                          cursorPosition: aElement.selectionStart,
+                                                          anchorPosition: aElement.selectionEnd}));
+    } catch (e) {
+      Logger.warn("InputMethodHandler: sending async message failed", e);
+    }
+  },
+
+  _resetInputContext: function(aElement) {
+    try {
+      let winId = Services.embedlite.getIDByWindow(aElement.ownerGlobal);
+      Services.embedlite.sendAsyncMessage(winId, "InputMethodHandler:ResetInputContext", "[]");
+    } catch (e) {
+      Logger.warn("InputMethodHandler: sending async message failed", e);
+    }
+  },
+
+  // We only want to show login suggestions for certain elements
+  _isAutoComplete: function(aElement) {
```

Comment:
Smells like a copy-paste from FormAssistant.js. I'd recommend different kind of composition.

There are two options I think

- Add the beef of this one to the FormAssistant.js. Question then being that we should that FormAssistant.js be called?
- Alternatively we could think of splitting FormAssistant into two and use common part from both FormAssistant.js and InputMethodHandler.js.

I do not have a strong take yet which is better way as both could potentially be good.

Reply:
Yep, this is copy-paste, but the implementation of _isAutoComplete is slightly different from the one in FormAssistant.js. About comment line, I missed it. It should be changed or removed. The second is preferable in this context I think.

Reply:
Alright. Created a follow up task to sync/align copy-pasted parts (JB#55725) - I think we can live with this for time being.